src/net/ossl.cc

```diff
@@ -0,0 +1,2073 @@
+        if (res == 1) {
+            return wait_for_output();
+        } else if (res == 0) {
+            return yield().then([this] { return do_shutdown(); });
+        } else {
+            auto ssl_err = SSL_get_error(_ssl.get(), res);
+            switch (ssl_err) {
+            case SSL_ERROR_NONE:
+                // this is weird, yield and try again
+                return yield().then([this] { return do_shutdown(); });
+            case SSL_ERROR_ZERO_RETURN:
+                // Looks like the other end is done, so let's just assume we're
+                // done as well
+                return wait_for_output();
+            case SSL_ERROR_WANT_READ:
+            case SSL_ERROR_WANT_WRITE:
+                return wait_for_output().then([this, ssl_err] {
+                    // In neither case do we actually want to pull data off of the socket (yet)
+                    // If we initiate the shutdown, then we just send the shutdown alert and wait
+                    // for EOF (outside of this function)
+                    if (ssl_err == SSL_ERROR_WANT_READ) {
+                        return make_ready_future();
+                    } else {
+                        return do_shutdown();
+                    }
+                });
+            case SSL_ERROR_SYSCALL:
+            {
+                auto err = make_ossl_error("System error during shutdown");
+                return handle_output_error(std::move(err));
+            }
+            case SSL_ERROR_SSL:
+            {
+                if (ERR_GET_REASON(ERR_peek_error()) == SSL_R_APPLICATION_DATA_AFTER_CLOSE_NOTIFY) {
+                    // This may have resulted in a race condition where we receive a packet immediately after
+                    // sending out the close notify alert.  In this situation, retry shutdown silently
+                    ERR_clear_error();
+                    return yield().then([this] { return do_shutdown(); });
+                }
+                auto err = make_ossl_error("Error occurred during SSL shutdown");
+                return handle_output_error(std::move(err));
+            }
+            default:
+            {
+                auto err = std::runtime_error(
+                  "Unknown error occurred during SSL shutdown");
+                return handle_output_error(std::move(err));
+            }
+            }
+        }
+    }
+
+    void verify() {
+        if (!_creds->_enable_certificate_verification) {
+            return;
+        }
+        // A success return code (0) does not signify if a cert was presented or not, that
+        // must be explicitly queried via SSL_get_peer_certificate
+        auto res = SSL_get_verify_result(_ssl.get());
+        if (res != X509_V_OK) {
+            auto stat_str(X509_verify_cert_error_string(res));
+            auto dn = extract_dn_information();
+            if (dn) {
+                std::string_view stat_str_view{stat_str};
+                if (stat_str_view.ends_with(" ")) {
```

Comment:
Nit: I assume this is an openssl quirk of some sort? Maybe a while instead (instead of assuming just one space), or maybe even a proper back and front whitespace strip?